x-pack/legacy/plugins/uptime/public/components/functional/monitor_list/monitor_list_page_size_select.tsx

```diff
@@ -0,0 +1,132 @@
+  <EuiButtonEmpty
+    color="text"
+    data-test-subj="xpack.uptime.monitorList.pageSizeSelect.popoverOpen"
+    iconType="arrowDown"
+    iconSide="right"
+    onClick={() => setIsOpen(true)}
+  >
+    <FormattedMessage
+      id="xpack.uptime.monitorList.pageSizePopoverButtonText"
+      defaultMessage="Rows per page: {size}"
+      values={{ size }}
+    />
+  </EuiButtonEmpty>
+);
+
+interface ContextItemProps {
+  'data-test-subj': string;
+  key: string;
+  numRows: number;
+}
+
+const items: ContextItemProps[] = [
+  {
+    'data-test-subj': 'xpack.uptime.monitorList.pageSizeSelect.sizeSelectItem10',
+    key: '10 rows',
+    numRows: 10,
+  },
+  {
+    'data-test-subj': 'xpack.uptime.monitorList.pageSizeSelect.sizeSelectItem25',
+    key: '25 rows',
+    numRows: 25,
+  },
+  {
+    'data-test-subj': 'xpack.uptime.monitorList.pageSizeSelect.sizeSelectItem50',
+    key: '50 rows',
+    numRows: 50,
+  },
+  {
+    'data-test-subj': 'xpack.uptime.monitorList.pageSizeSelect.sizeSelectItem100',
+    key: '100 rows',
+    numRows: 100,
+  },
+];
+
+const LOCAL_STORAGE_KEY = 'xpack.uptime.monitorList.pageSize';
+
+interface MonitorListPageSizeSelectProps {
+  size: number;
+  setSize: (value: number) => void;
+}
+
+/**
+ * This component wraps the underlying UI functionality to make the component more testable.
+ * The features leveraged in this function are tested elsewhere, and are not novel to this component.
+ */
+export const MonitorListPageSizeSelect: React.FC<MonitorListPageSizeSelectProps> = ({
+  size,
+  setSize,
+}) => {
+  const [, setUrlParams] = useUrlParams();
+
+  useEffect(() => {
+    localStorage.setItem(LOCAL_STORAGE_KEY, size.toString());
+  }, [size]);
+
+  return (
+    <MonitorListPageSizeSelectComponent size={size} setSize={setSize} setUrlParams={setUrlParams} />
+  );
+};
+
+interface ComponentProps extends MonitorListPageSizeSelectProps {
+  setUrlParams: UpdateUrlParams;
+}
+
+/**
+ * This function contains the UI functionality for the page select feature. It's agnostic to any
+ * external services/features, and focuses only on providing the UI and handling user interaction.
+ */
+export const MonitorListPageSizeSelectComponent: React.FC<ComponentProps> = ({
+  size,
+  setSize,
+  setUrlParams,
+}) => {
+  const [isOpen, setIsOpen] = useState(false);
+  return (
+    <EuiPopover
+      button={<PopoverButton setIsOpen={value => setIsOpen(value)} size={size} />}
+      isOpen={isOpen}
+      closePopover={() => setIsOpen(false)}
+      anchorPosition="upLeft"
+    >
+      <EuiContextMenuPanel
+        items={items.map(({ 'data-test-subj': dataTestSubj, key, numRows }) => (
+          <EuiContextMenuItem
+            data-test-subj={dataTestSubj}
+            key={key}
+            icon={size === numRows ? 'check' : 'empty'}
+            onClick={() => {
+              setSize(numRows);
+              // reset pagination because the page size has changed
```

Comment:
I don't understand why we need to reset the pagination due to a size change.

Reply:
I had some wonky behavior without resetting pagination on size change. I think it has to do with the current page ID governing the next/prev pages.

Basically, it became possible for pages to have `< size` rows, even though there were `> size` monitors configured.

I reasoned that since size is likely something you set and forget it wouldn't impact the UX very much to reset pagination when the size updates.

Reply:
Ahhhh, makes sense!